### PR TITLE
splitted the numpy backend into several files.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
### Summary

The backend files are quite big:
`theano_backend.py` : 2900 lines
`tensorflow_backend.py`: 4600 lines
`cntk_backend.py`: 2700 lines.

They are big enough to be splitted into several files which should help in making those files more readable and accessible. 

Furthermore in `keras.backend.__init__.py` we do things like `from .theano_backend import *` which are polluting the local namespace with hidden functions, functions shadowing python built-ins (`all`, `any`...) and many imports (how many times have we seen people do `K.tf` ?).
So we would benefit from a `__init__.py` where we export public definitions.

I've also put all functions shadowing python built-in names into a separate file. Allowing us to remove the hacky:

```python
py_all = all
py_sum = sum
...
```
For the other implementations, we won't need to use `py_slice` anymore.

### PR Overview

To start, I implement this splitting with the numpy backend which is the simplest and the smallest of all so that we can easily discuss what goes where.

Here is a summary of what moved where.

```python
# control_flow.py
rnn
switch
learning_phase
set_learning_phase
in_train_phase
in_test_phase

# convolutions.py
conv1d
conv2d
conv3d
depthwise_conv2d
separable_conv1d
separable_conv2d
conv2d_transpose
conv3d_transpose
pool2d
pool3d
bias_add

# ctc.py
ctc_decode

# elementwise.py
prod
cumsum
cumprod
mean
std
var
argmax
argmin
sqrt
logsumexp
clip
equal
not_equal
greater
greater_equal
less
less_equal
maximum
minimum
square
exp
log
sign
cos
sin

# linear_algebra.py
dot
batch_dot
transpose
gather

# nn.py
elu
relu
softmax
softplus
categorical_crossentropy
binary_crossentropy
sigmoid
hard_sigmoid
tanh
dropout
l2_normalize

# overrides.py
all
any
sum
max
min
abs
round
pow

# shape_operations.py
int_shape
ndim
get_variable_shape
concatenate
reshape
permute_dimensions
repeat_elements
resize_images
resize_volumes
repeat
arange
tile
flatten
batch_flatten
temporal_padding
spatial_2d_padding
spatial_3d_padding
one_hot
reverse
expand_dims
squeeze

# value_manipulations.py
get_value
print_tensor

# variable_manipulation.py
variable
constant
dtype
eval
zeros
zeros_like
ones
ones_like
eye
random_uniform_variable
random_normal_variable
count_params
```

I took the names of the files from the names of the sections in `theano_backend.py`. Since I'm really bad at naming thing, don't feel bad to suggest to change them all.

- [ ] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

File: tests/keras/backend/backend_test.py [hunk 1]
```diff
@@ -1281,8 +1281,8 @@ def test_separable_conv(self,
                                           (input_depth, depth_multiplier))
         _, pointwise = parse_shape_or_val((1,) * len(kernel_shape) +
                                           (input_depth * depth_multiplier, 7))
-        y1 = KNP.separable_conv(x, depthwise, pointwise,
-                                padding=padding, data_format=data_format)
+        y1 = getattr(KNP, op)(x, depthwise, pointwise,
```
Comment (farizrahman4u) [inline tests/keras/backend/backend_test.py]: Is this related to this PR?

**Comment by farizrahman4u** [comment]
`eval` is also a builtin :)

Though I am not sure about the filename `overrides.py`. Maybe they should go in their respective op files.
They fact that they override builtins is a technical detail but its better to classify ops just by their behavior.

**Comment by gabrieldemarmiesse** [comment]
> `eval` is also a builtin

Very true.

> Though I am not sure about the filename overrides.py. Maybe they should go in their respective op files. They fact that they override builtins is a technical detail but its better to classify ops just by their behavior.

We have:
* 6 usages of `py_any`
* 5 usages of `py_all`
* 1 usage of `py_sum`
* 31 usages of `py_slice`

Readability counts. If a user wants to understand what a function does and faces a `py_slice`, this user will have to read the code and understand that there is a hack going on to be able to understand the function which he was interested in. It is also error prone for people wanting to contribute to the codebase. They will also have to understand that they can't use common python keywords anymore.

But I also agree that this `override.py` is ugly because people will be surprise that they won't find what they are looking for in the right files.

I have been thinking about this, and what do you think about the following:

* We delete `override.py`.
* We put the functions in the right files.
* We do 
```python
def keras_abs():
    ....
```
* In the `__init__.py`, we do 
```python
from .elementwise import keras_abs as abs
```

People who use ctrl+f will still find what they are looking for, we don't shadow built-ins and the declarations are in the right files. I'm happy, you're happy, the readers are happy, the contributors are happy and everyone lived happily ever after.

File: tests/keras/backend/backend_test.py (hunk 1 above)
Comment (gabrieldemarmiesse) [reply tests/keras/backend/backend_test.py]: It is and for a very nice reason. Since we don't pollute the backend namespace, you can't use K.separable_conv, because it's not supposed to be public. Similarly to how people used to do `K.tf`. This change is necessary for the build to pass.

**Comment by farizrahman4u** [comment]
> 
    6 usages of py_any
    5 usages of py_all
    1 usage of py_sum
    31 usages of py_slice

These usages are in other backends and will still be there after this PR is merged, right? Am I missing something.

FWIW: backing up builtins is a common practice. See https://github.com/numpy/numpy/blob/master/numpy/core/fromnumeric.py#L1969


> If a user wants to understand what a function does and faces a py_slice

Here "user" refers to someone browsing through codebase/ trying to contribute correct? Or a normal Keras user?

Users are supposed to do `import keras.backend as K`, not `from keras.backend import *`.

**Comment by gabrieldemarmiesse** [comment]
@farizrahman4u . I removed `overrides.py` like you suggested. You can see what went where in the `__init__.py`. Actually everything went in `elementwise.py`. Could you review again?

**Comment by farizrahman4u** [comment]
@gabrieldemarmiesse Even though there are no public API changes, this PR would require @fchollet's approval due to its nature.

**Comment by gabrieldemarmiesse** [comment]
Okay, sure, that makes sense.

**Comment by gabrieldemarmiesse** [comment]
I'll close this issue and make a new one reworked later.
